src/test/scala/com/mozilla/telemetry/streaming/TestErrorAggregator.scala

```diff
@@ -386,6 +424,89 @@ class TestErrorAggregator extends FlatSpec with Matchers with BeforeAndAfterAll
     val df = ErrorAggregator.aggregate(spark.sqlContext.createDataset(messages).toDF, raiseOnError = false, online = false)
 
     df.where("application <> 'Firefox'").count() should be (0)
+  }
 
+  "the aggregator" should "correctly read from kafka" taggedAs(DockerComposeTag) in {
+    val conf = new Properties()
+    conf.put("bootstrap.servers", kafkaBrokers)
+    conf.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer")
+    conf.put("value.serializer", "org.apache.kafka.common.serialization.ByteArraySerializer")
+
+    val kafkaProducer = new KafkaProducer[String, Array[Byte]](conf)
+    createTopic(ErrorAggregator.kafkaTopic, kafkaTopicPartitions)
+
+    // TODO: Test out different wait times (should not effect Spark)
+    def waitSend(rs: Seq[Array[Byte]], wait: Int = 100): Unit = {
+      rs.foreach{ v =>
+        val record = new ProducerRecord[String, Array[Byte]](ErrorAggregator.kafkaTopic, v)
+        kafkaProducer.send(record)
+        java.lang.Thread.sleep(wait)
+      }
+    }
+
+    val earlier = (TestUtils.generateMainMessages(k, timestamp=Some(1459994516000000000L)) ++
+      TestUtils.generateCrashMessages(k, timestamp=Some(1459994516000000000L))).map(_.toByteArray)
+
+    val later = (TestUtils.generateMainMessages(k, timestamp=Some(1459996516000000000L)) ++
+      TestUtils.generateCrashMessages(k, timestamp=Some(1459996516000000000L))).map(_.toByteArray)
+
+    val expectedTotal = 4 * k
+
+    // Spark Bug?
+    // Send earlier, wait 30 secs, send later
+    //
+    // Currently this is the only way to get spark
+    // to actually write the files (lowering to 20
+    // seconds makes it fail sometimes, and lowering
+    // it to 10 seconds makes it fail every time)
+    def sendRecords(): Unit = {
+      waitSend(earlier)
+      java.lang.Thread.sleep(30000)
+      waitSend(later)
+    }
+
+    val listener = new StreamingQueryListener {
+      var messagesSeen = 0L
+
+      override def onQueryStarted(event: StreamingQueryListener.QueryStartedEvent): Unit = {
+        sendRecords()
+      }
+
+      override def onQueryProgress(event: StreamingQueryListener.QueryProgressEvent): Unit = {
+        messagesSeen += event.progress.numInputRows
+
+        // Spark Bug? Sending messages after getting QueryStartedEvent are still missed
+        // Sending again if the query hasn't seen any
+        if(messagesSeen == 0){
```

Comment:
So this function should run on every trigger, right? That means we keep sending it at the same interval of the trigger, but it takes at least 30 seconds to finish. Unless this is run synchronously, but I don't think so.